lib/router_constraints.rb

```diff
@@ -58,28 +62,53 @@ def matches?(request, path = nil)
       unescape_uri(request.original_fullpath)
     # retrieves the hierarchy and member portions of loc/id's
     hierarchy_member = path.split('?', 2).first.split('///', 2).first
-    element = @find_in_klass.find_with_locid(hierarchy_member)
-    ontology = element.respond_to?(:ontology) ? element.ontology : element
-    result = !ontology.nil?
+    element =
+      elements_cache[hierarchy_member] ||
+      LocIdBaseModel.find_with_locid(hierarchy_member)
+
+    if element.is_a?(@find_in_klass)
+      elements_cache.delete(hierarchy_member)
+      clear_elements_cache if elements_cache_clearing_scheduled?
+      assign_path_parameters(request, element)
+
+      true
+    else
+      if !elements_cache.key?(hierarchy_member) && element
+        elements_cache[hierarchy_member] = element
+      end
+      false
+    end
+  end
 
-    if result
-      proof_attempt = element.proof_attempt if @map[:proof_attempt]
-      theorem = element.theorem if @map[:theorem]
-      path_params =
-        if ontology.respond_to?(:repository)
-          {repository_id: ontology.repository.to_param}
-        else
-          {}
-        end
-      path_params[@map[:proof_attempt]] = proof_attempt.id if @map[:proof_attempt]
-      path_params[@map[:theorem]] = theorem.id if @map[:theorem]
-      path_params[@map[:ontology]] = ontology.id if @map[:ontology]
-      path_params[@map[:element]] = element.id if @map[:element]
-
-      add_path_parameters(request, path_params)
+  def elements_cache
+    self.class.instance_variable_get(:@elements_cache)
+  end
+
+  # To keep the memory footprint low in case of unproper removal of cached
+  # elements, we clear the cache periodically. This prevents memory leaks.
+  def clear_elements_cache
+    next_clearing = Time.now + ELEMENTS_CACHE_CLEARING_INTERVAL
+    self.class.instance_variable_set(:@elements_cache_clear_at, next_clearing)
+
+    self.class.instance_variable_set(:@elements_cache, {})
+  end
+
+  def elements_cache_clearing_scheduled?
+    Time.now > self.class.instance_variable_get(:@elements_cache_clear_at)
+  end
+
+  def assign_path_parameters(request, element)
+    ontology = element.is_a?(Ontology) ? element : element.ontology
+
+    proof_attempt = element.proof_attempt if @map[:proof_attempt]
+    theorem = element.theorem if @map[:theorem]
+
+    path_params = {repository_id: ontology.repository.to_param}
+    %i(proof_attempt theorem ontology element).each do |param|
+      path_params[@map[param]] = eval(param.to_s).id if @map[param]
```

Comment:
It isn't in this case because we only use meta programming to access the variable. I'm not sure, though, if this code gives us any benefit (less readable, only a little more compact, a lot more DRY). @ebolloff and @phyrog, please comment on this.